In [164]:
# Import necessary libraries
import pandas as pd
from gurobipy import Model, GRB

# Load datasets
farms = pd.read_csv('/Users/Sam/Downloads/farms.csv')
processing = pd.read_csv('/Users/Sam/Downloads/processing.csv')
centers = pd.read_csv('/Users/Sam/Downloads/centers.csv')

# Define dimensions
num_farms = len(farms)
num_plants = len(processing)
num_centers = len(centers)

# Extract transportation cost data
transport_cost_farm_to_plant = [[farms.iloc[i, 4+j] for j in range(num_plants)] for i in range(num_farms)]
transport_cost_plant_to_center = [[processing.iloc[j, 4+k] for k in range(num_centers)] for j in range(num_plants)]

# Function to initialize a new model
def initialize_model():
    # Create a new model
    model = Model('BioAgri Optimization')

    # Create decision variables
    x = model.addVars(num_farms, num_plants, vtype=GRB.CONTINUOUS, name="x")  # Raw material transport
    y = model.addVars(num_plants, num_centers, vtype=GRB.CONTINUOUS, name="y")  # Fertilizer transport

    # Objective function: Minimize total cost
    model.setObjective(
        sum(x[i, j] * (farms.loc[i, 'Cost_Per_Ton'] + transport_cost_farm_to_plant[i][j])
            for i in range(num_farms) for j in range(num_plants)) +
        sum(y[j, k] * (processing.loc[j, 'Processing_Cost_Per_Ton'] + transport_cost_plant_to_center[j][k])
            for j in range(num_plants) for k in range(num_centers)),
        GRB.MINIMIZE
    )

    return model, x, y


In [165]:
# Initialize a new model
model, x, y = initialize_model()

# Add basic constraints
for i in range(num_farms):
    model.addConstr(sum(x[i, j] for j in range(num_plants)) <= farms.loc[i, 'Bio_Material_Capacity_Tons'], 
                    name=f"FarmSupply_{i}")
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) <= processing.loc[j, 'Capacity_Tons'], 
                    name=f"PlantCapacity_{j}")
for k in range(num_centers):
    model.addConstr(sum(y[j, k] for j in range(num_plants)) == centers.loc[k, 'Requested_Demand_Tons'], 
                    name=f"CenterDemand_{k}")
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) == sum(y[j, k] for k in range(num_centers)), 
                    name=f"FlowBalance_{j}")

# Optimize the model
model.optimize()

# Output results for (a) and (b)
if model.status == GRB.OPTIMAL:
    print("Optimal Cost:", model.objVal)
else:
    print("No optimal solution found.")


Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 387 rows, 6318 columns and 17118 nonzeros
Model fingerprint: 0x3aa78cc0
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [6e+00, 3e+02]
  Bounds range     [0e+00, 0e+00]
  RHS range        [6e+01, 3e+04]
Presolve time: 0.00s
Presolved: 387 rows, 6318 columns, 17118 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.9457439e+05   2.902000e+04   0.000000e+00      0s
     330    2.2970900e+06   0.000000e+00   0.000000e+00      0s

Solved in 330 iterations and 0.01 seconds (0.02 work units)
Optimal objective  2.297089973e+06
Optimal Cost: 2297089.972722625


* c

In [166]:
# Initialize a new model for part (c)
model, x, y = initialize_model()

# Add basic constraints
for i in range(num_farms):
    model.addConstr(sum(x[i, j] for j in range(num_plants)) <= farms.loc[i, 'Bio_Material_Capacity_Tons'], 
                    name=f"FarmSupply_{i}")
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) <= processing.loc[j, 'Capacity_Tons'], 
                    name=f"PlantCapacity_{j}")
for k in range(num_centers):
    model.addConstr(sum(y[j, k] for j in range(num_plants)) == centers.loc[k, 'Requested_Demand_Tons'], 
                    name=f"CenterDemand_{k}")
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) == sum(y[j, k] for k in range(num_centers)), 
                    name=f"FlowBalance_{j}")

# Add regional restriction constraints
for j in range(num_plants):
    plant_region = processing.loc[j, 'Region']
    for k in range(num_centers):
        center_region = centers.loc[k, 'Region']
        if plant_region != center_region:
            model.addConstr(y[j, k] == 0, name=f"RegionalRestriction_Plant{j}_Center{k}")

# Optimize the model
model.optimize()

# Output results for (c)
if model.status == GRB.OPTIMAL:
    print("Optimal Cost (Regional Restriction):", model.objVal)
else:
    print("No optimal solution found.")


Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 1761 rows, 6318 columns and 18492 nonzeros
Model fingerprint: 0x9c296277
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [6e+00, 3e+02]
  Bounds range     [0e+00, 0e+00]
  RHS range        [6e+01, 3e+04]
Presolve removed 1374 rows and 1374 columns
Presolve time: 0.00s
Presolved: 387 rows, 4944 columns, 14370 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    2.2479485e+05   3.627500e+03   0.000000e+00      0s
     320    2.3234885e+06   0.000000e+00   0.000000e+00      0s

Solved in 320 iterations and 0.01 seconds (0.01 work units)
Optimal objective  2.323488497e+06
Optimal Cost (Regional Restriction): 2323488.4970226064


* d

In [167]:
# Initialize a new model for part (d)
model, x, y = initialize_model()

# Add basic constraints
for i in range(num_farms):
    model.addConstr(sum(x[i, j] for j in range(num_plants)) <= farms.loc[i, 'Bio_Material_Capacity_Tons'], 
                    name=f"FarmSupply_{i}")
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) <= processing.loc[j, 'Capacity_Tons'], 
                    name=f"PlantCapacity_{j}")
for k in range(num_centers):
    model.addConstr(sum(y[j, k] for j in range(num_plants)) == centers.loc[k, 'Requested_Demand_Tons'], 
                    name=f"CenterDemand_{k}")
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) == sum(y[j, k] for k in range(num_centers)), 
                    name=f"FlowBalance_{j}")

# Add quality restriction constraints
for i in range(num_farms):
    if farms.loc[i, 'Quality'] < 3:
        for j in range(num_plants):
            model.addConstr(x[i, j] == 0, name=f"QualityRestriction_Farm{i}_Plant{j}")

# Optimize the model
model.optimize()

# Output results for (d)
if model.status == GRB.OPTIMAL:
    print("Optimal Cost (Highest Quality Only):", model.objVal)
else:
    print("No optimal solution found.")


Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 3519 rows, 6318 columns and 20250 nonzeros
Model fingerprint: 0xa45bebf3
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [6e+00, 3e+02]
  Bounds range     [0e+00, 0e+00]
  RHS range        [6e+01, 3e+04]
Presolve removed 3306 rows and 3132 columns
Presolve time: 0.00s
Presolved: 213 rows, 3186 columns, 7722 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.9457439e+05   3.627500e+03   0.000000e+00      0s
     253    5.7129767e+06   0.000000e+00   0.000000e+00      0s

Solved in 253 iterations and 0.01 seconds (0.01 work units)
Optimal objective  5.712976673e+06
Optimal Cost (Highest Quality Only): 5712976.672909363


* e

In [168]:
# Initialize a new model for part (e) - Sourcing Risk Mitigation
model, x, y = initialize_model()

# Add basic constraints
for i in range(num_farms):
    model.addConstr(sum(x[i, j] for j in range(num_plants)) <= farms.loc[i, 'Bio_Material_Capacity_Tons'], 
                    name=f"FarmSupply_{i}")
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) <= processing.loc[j, 'Capacity_Tons'], 
                    name=f"PlantCapacity_{j}")
for k in range(num_centers):
    model.addConstr(sum(y[j, k] for j in range(num_plants)) == centers.loc[k, 'Requested_Demand_Tons'], 
                    name=f"CenterDemand_{k}")
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) == sum(y[j, k] for k in range(num_centers)), 
                    name=f"FlowBalance_{j}")

# Add sourcing risk mitigation constraint
total_supply = farms['Bio_Material_Capacity_Tons'].sum()
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) <= 0.03 * total_supply, 
                    name=f"SourcingRisk_Plant{j}")

# Optimize the model
model.optimize()

# Output results for (e) Part 1
if model.status == GRB.OPTIMAL:
    print("Optimal Cost (Sourcing Risk Mitigation):", model.objVal)
else:
    print("No optimal solution found.")


Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 405 rows, 6318 columns and 21600 nonzeros
Model fingerprint: 0x52a9cb0c
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [6e+00, 3e+02]
  Bounds range     [0e+00, 0e+00]
  RHS range        [6e+01, 3e+04]
Presolve removed 18 rows and 0 columns
Presolve time: 0.00s
Presolved: 387 rows, 6318 columns, 17118 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.9457439e+05   3.627500e+03   0.000000e+00      0s
     408    2.3117826e+06   0.000000e+00   0.000000e+00      0s

Solved in 408 iterations and 0.01 seconds (0.02 work units)
Optimal objective  2.311782560e+06
Optimal Cost (Sourcing Risk Mitigation): 2311782.559847675


In [169]:
# Initialize a new model for part (e) - Supply Risk Mitigation
model, x, y = initialize_model()

# Add basic constraints
for i in range(num_farms):
    model.addConstr(sum(x[i, j] for j in range(num_plants)) <= farms.loc[i, 'Bio_Material_Capacity_Tons'], 
                    name=f"FarmSupply_{i}")
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) <= processing.loc[j, 'Capacity_Tons'], 
                    name=f"PlantCapacity_{j}")
for k in range(num_centers):
    model.addConstr(sum(y[j, k] for j in range(num_plants)) == centers.loc[k, 'Requested_Demand_Tons'], 
                    name=f"CenterDemand_{k}")
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) == sum(y[j, k] for k in range(num_centers)), 
                    name=f"FlowBalance_{j}")

# Add supply risk mitigation constraint
for j in range(num_plants):
    for k in range(num_centers):
        model.addConstr(y[j, k] <= 0.5 * centers.loc[k, 'Requested_Demand_Tons'], 
                        name=f"SupplyRisk_Plant{j}_Center{k}")

# Optimize the model
model.optimize()

# Output results for (e) Part 2
if model.status == GRB.OPTIMAL:
    print("Optimal Cost (Supply Risk Mitigation):", model.objVal)
else:
    print("No optimal solution found.")


Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 2223 rows, 6318 columns and 18954 nonzeros
Model fingerprint: 0x3179cd01
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [6e+00, 3e+02]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+01, 3e+04]
Presolve removed 1836 rows and 0 columns
Presolve time: 0.00s
Presolved: 387 rows, 6318 columns, 17118 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.9457439e+05   5.441250e+03   0.000000e+00      0s
     398    2.3019498e+06   0.000000e+00   0.000000e+00      0s

Solved in 398 iterations and 0.01 seconds (0.02 work units)
Optimal objective  2.301949785e+06
Optimal Cost (Supply Risk Mitigation): 2301949.784839987


* f

Here’s how to answer question (f) in English, including an analysis of all four scenarios from (c), (d), and the two parts of (e), followed by the combined implementation code.

---

### **Question (f): Explanation and Analysis**

To determine which options are financially defensible, we analyze the results from (c), (d), and (e) in terms of cost impact, operational feasibility, and alignment with strategic goals.

#### **Scenario (c): Regional Restriction**
- **Result**: Optimal cost = **2,323,488.50**.
- **Explanation**: Restricting processing plants to serve home centers only within the same region aligns with practical logistical considerations. While this constraint slightly increases costs compared to a fully flexible model, it reflects realistic limitations, such as transportation efficiency and regulatory boundaries.
- **Conclusion**: **Financially defensible**, as it balances operational feasibility with a manageable cost increase.

#### **Scenario (d): High-Quality Raw Material**
- **Result**: Optimal cost = **5,712,976.67**.
- **Explanation**: Sourcing only high-quality raw materials (quality 3 and 4) nearly doubles the cost compared to (c). While this may improve product quality, the significant cost increase is unlikely to yield proportional benefits in revenue or market competitiveness.
- **Conclusion**: **Not financially defensible**, as the cost increase is excessively high and does not justify the potential benefits.

#### **Scenario (e-1): Sourcing Risk Mitigation**
- **Result**: Optimal cost = **2,311,782.56**.
- **Explanation**: Limiting each facility to processing no more than 3% of the total raw material reduces the risk of over-reliance on specific facilities. The impact on cost is minimal (slightly lower than (c)), demonstrating that this constraint improves risk management without significantly affecting financial performance.
- **Conclusion**: **Financially defensible**, as it enhances supply chain resilience at a negligible cost.

#### **Scenario (e-2): Supply Risk Mitigation**
- **Result**: Optimal cost = **2,301,949.78**.
- **Explanation**: Restricting a processing facility from supplying more than 50% of a single home center’s demand addresses dependency risk. This constraint reduces cost further compared to (c) and (e-1), indicating an improvement in logistical efficiency and risk distribution.
- **Conclusion**: **Financially defensible**, as it reduces supply chain risk while optimizing costs.

---

### **Implementation Plan for All Defensible Options**
Combining all defensible options ((c), (e-1), and (e-2)) offers the best trade-off between cost and risk mitigation. While these constraints introduce operational complexity, they enhance the robustness and sustainability of the supply chain network.

#### **Why This Combination is Financially Defensible**
- Balances cost optimization and risk reduction.
- Reflects realistic logistical and operational limitations.
- Enhances resilience against supply chain disruptions without incurring excessive costs.

---




In [170]:
# Initialize a new model for part (f)
model, x, y = initialize_model()

# Add basic constraints
for i in range(num_farms):
    model.addConstr(sum(x[i, j] for j in range(num_plants)) <= farms.loc[i, 'Bio_Material_Capacity_Tons'], 
                    name=f"FarmSupply_{i}")
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) <= processing.loc[j, 'Capacity_Tons'], 
                    name=f"PlantCapacity_{j}")
for k in range(num_centers):
    model.addConstr(sum(y[j, k] for j in range(num_plants)) == centers.loc[k, 'Requested_Demand_Tons'], 
                    name=f"CenterDemand_{k}")
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) == sum(y[j, k] for k in range(num_centers)), 
                    name=f"FlowBalance_{j}")

# Add regional restriction constraints
for j in range(num_plants):
    plant_region = processing.loc[j, 'Region']
    for k in range(num_centers):
        center_region = centers.loc[k, 'Region']
        if plant_region != center_region:
            model.addConstr(y[j, k] == 0, name=f"RegionalRestriction_Plant{j}_Center{k}")
        
# Add sourcing risk mitigation constraint
total_supply = farms['Bio_Material_Capacity_Tons'].sum()
for j in range(num_plants):
    model.addConstr(sum(x[i, j] for i in range(num_farms)) <= 0.03 * total_supply, 
                    name=f"SourcingRisk_Plant{j}")

# Add supply risk mitigation constraint
for j in range(num_plants):
    for k in range(num_centers):
        model.addConstr(y[j, k] <= 0.5 * centers.loc[k, 'Requested_Demand_Tons'], 
                        name=f"SupplyRisk_Plant{j}_Center{k}")

# Optimize the model
model.optimize()

# Output results for (f)
if model.status == GRB.OPTIMAL:
    print("Optimal Cost (All Defensible Options):", model.objVal)
else:
    print("No optimal solution found with all defensible options.")


Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 3615 rows, 6318 columns and 24810 nonzeros
Model fingerprint: 0xb36e5572
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [6e+00, 3e+02]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+01, 3e+04]
Presolve removed 3228 rows and 1374 columns
Presolve time: 0.00s
Presolved: 387 rows, 4944 columns, 14370 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    2.2479485e+05   5.441250e+03   0.000000e+00      0s
     397    2.3372934e+06   0.000000e+00   0.000000e+00      0s

Solved in 397 iterations and 0.01 seconds (0.02 work units)
Optimal objective  2.337293375e+06
Optimal Cost (All Defensible Options): 2337293.3745054523


---

### **Expected Outcome**
The result of this combined model will provide the optimal cost when all financially defensible options are implemented together. If needed, the constraints can be further fine-tuned based on business priorities. Let me know if you need additional insights or refinements!

* g

### **Answer for Question (g)**

**Defending the Implementation of All Defensible Options**

Implementing all the defensible options increases the cost compared to the original system, but it represents a strong business decision due to the following key reasons:

1. **Risk Mitigation and Supply Chain Resilience**:
   - **Sourcing Risk Mitigation (e-1)** ensures no single facility becomes a bottleneck, improving the resilience of the supply chain. The cost impact is minimal (Optimal Cost: **2,311,782.56**).
   - **Supply Risk Mitigation (e-2)** prevents over-reliance on specific facilities for fulfilling home center demands, leading to improved distribution (Optimal Cost: **2,301,949.78**).

2. **Operational Feasibility**:
   - **Regional Restrictions (c)** reflect logistical realities, ensuring that processing plants serve nearby home centers, which aligns with real-world transportation limitations (Optimal Cost: **2,323,488.50**).

3. **Strategic Cost Control**:
   - The combined defensible options (f) result in an optimal cost of **2,337,293.37**, which is only a 0.6% increase over (c). This small cost increment is justified by the enhanced robustness and risk mitigation achieved.

4. **Avoiding Unsustainable Cost Increases**:
   - The high-quality raw material constraint (d) results in a significant cost increase to **5,712,976.67** (+145% vs. (c)), which is financially indefensible compared to the combined defensible options.

**Conclusion**:
While the combined defensible options increase costs slightly, they significantly enhance the robustness, feasibility, and sustainability of the supply chain. This approach ensures long-term resilience and operational efficiency, justifying the additional cost.

* h

In [171]:
# Function to test feasibility at different sourcing risk mitigation percentages
def test_sourcing_risk_threshold(starting_percentage, step_size):
    percentage = starting_percentage
    while percentage > 0:
        # Initialize a new model for each percentage
        model, x, y = initialize_model()

        # Add basic constraints
        for i in range(num_farms):
            model.addConstr(sum(x[i, j] for j in range(num_plants)) <= farms.loc[i, 'Bio_Material_Capacity_Tons'], 
                            name=f"FarmSupply_{i}")
        for j in range(num_plants):
            model.addConstr(sum(x[i, j] for i in range(num_farms)) <= processing.loc[j, 'Capacity_Tons'], 
                            name=f"PlantCapacity_{j}")
        for k in range(num_centers):
            model.addConstr(sum(y[j, k] for j in range(num_plants)) == centers.loc[k, 'Requested_Demand_Tons'], 
                            name=f"CenterDemand_{k}")
        for j in range(num_plants):
            model.addConstr(sum(x[i, j] for i in range(num_farms)) == sum(y[j, k] for k in range(num_centers)), 
                            name=f"FlowBalance_{j}")

        # Add regional restriction constraints
        for j in range(num_plants):
            plant_region = processing.loc[j, 'Region']
            for k in range(num_centers):
                center_region = centers.loc[k, 'Region']
                if plant_region != center_region:
                    model.addConstr(y[j, k] == 0, name=f"RegionalRestriction_Plant{j}_Center{k}")

        # Add sourcing risk mitigation constraint with the current percentage
        total_supply = farms['Bio_Material_Capacity_Tons'].sum()
        for j in range(num_plants):
            model.addConstr(sum(x[i, j] for i in range(num_farms)) <= (percentage / 100) * total_supply, 
                            name=f"SourcingRisk_Plant{j}")

        # Add supply risk mitigation constraint
        for j in range(num_plants):
            for k in range(num_centers):
                model.addConstr(y[j, k] <= 0.5 * centers.loc[k, 'Requested_Demand_Tons'], 
                                name=f"SupplyRisk_Plant{j}_Center{k}")

        # Optimize the model
        model.optimize()

        # Check feasibility
        if model.status == GRB.INFEASIBLE:
            print(f"Infeasibility reached at sourcing risk mitigation percentage: {percentage:.1f}%")
            return percentage
        percentage -= step_size  # Reduce percentage by step size

    print("Model remained feasible for all tested percentages.")
    return 0  # If no infeasibility is found


In [172]:
# Test the threshold for sourcing risk mitigation
threshold = test_sourcing_risk_threshold(starting_percentage=3.0, step_size=0.1)
print(f"Threshold for sourcing risk mitigation: {threshold:.1f}%")


Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 3615 rows, 6318 columns and 24810 nonzeros
Model fingerprint: 0xb36e5572
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [6e+00, 3e+02]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+01, 3e+04]
Presolve removed 3228 rows and 1374 columns
Presolve time: 0.00s
Presolved: 387 rows, 4944 columns, 14370 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    2.2479485e+05   5.441250e+03   0.000000e+00      0s
     397    2.3372934e+06   0.000000e+00   0.000000e+00      0s

Solved in 397 iterations and 0.01 seconds (0.02 work units)
Optimal objective  2.337293375e+06
Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors

### **Answer for Question (h)**

#### **Key Findings**
1. **Threshold for Feasibility**:
   - The supply chain becomes infeasible when the sourcing risk mitigation limit drops below **2.4%**.
   - At **2.5%**, the model remains feasible with an optimal transportation and procurement cost of **$5,755,911.11**.

2. **Managerial Interpretation**:
   - **Capacity Limits**: The result highlights that the supply chain has limited flexibility to absorb tighter restrictions on sourcing. Dropping the limit below **2.5%** creates excessive reliance on a few facilities, making the network unsustainable.
   - **Risk-Resilience Trade-Off**: While reducing the sourcing risk mitigation percentage might seem desirable to lower dependency risks, the supply chain cannot operate below the threshold due to resource and capacity constraints.

3. **Implications for Supply Chain Management**:
   - **Balance Constraints**: Managers must carefully balance risk mitigation policies with operational feasibility to avoid disruption.
   - **Expand Capacity**: To implement stricter sourcing limits, additional investments in facility capacity or diversification of suppliers may be necessary.
   - **Monitor Critical Limits**: The **2.5% threshold** serves as a guide for future policy development, ensuring the supply chain remains robust while addressing risk.

4. **Conclusion**:
   - The supply chain’s limited capacity for risk mitigation emphasizes the need for strategic adjustments rather than imposing overly restrictive policies. Maintaining feasible limits is crucial for sustaining operations and achieving long-term goals.

* Question B

In [173]:
import pandas as pd
from gurobipy import Model, GRB

# Load the dataset
gym = pd.read_csv('/Users/Sam/Downloads/updated_gym_data.csv')

# Initialize the Gurobi model
model = Model("Weightlifting_Program")

# Number of exercises
n_exercises = gym.shape[0]

# Decision variables: Proportion of workout allocated to each exercise
# Variable names are generated dynamically based on exercise indices
x = model.addVars(n_exercises, lb=0, ub=1, name="x")

# Total allocation constraint: Sum of proportions must equal 1
model.addConstr(x.sum() == 1, "TotalAllocation")

# Print the number of decision variables
print(f"Number of decision variables: {n_exercises}")


Number of decision variables: 2637


# Decision Variables

Each exercise in the dataset represents a potential decision variable. Since the dataset contains 2,637 exercises, there are 2,637 decision variables. Each variable \(x_i\) corresponds to the proportion of the workout program allocated to exercise \(i\).


Output

Number of decision variables: 2,637.
Range of each decision variable: 
[
0
,
1
]
[0,1].


In [174]:
gym.head()

Exercise      Category    BodyPart Equipment  \
0      Bench Press With Short Bands  Powerlifting       Chest     Bands   
1                Hip Lift with Band  Powerlifting      Glutes     Bands   
2  Band Good Morning (Pull Through)  Powerlifting  Hamstrings     Bands   
3                   Speed Box Squat  Powerlifting  Quadriceps     Bands   
4            Partner plank band row      Strength  Abdominals     Bands   

     Difficulty  Stimulus-to-Fatigue  Expected Time  Hypertrophy Rating  
0      Beginner             0.817884      15.518089            0.596124  
1      Beginner             0.768902      14.655351            0.623237  
2      Beginner             0.792188      16.292358            0.601159  
3  Intermediate             0.599044      17.109781            0.800347  
4  Intermediate             0.730726      14.212727            0.461565

In [175]:
# import pandas as pd

# # Load the dataset
# gym_data_path = '/Users/Sam/Downloads/updated_gym_data.csv'  # Update this path if necessary
# gym = pd.read_csv(gym_data_path)

# # Normalize exercise names for matching
# gym['Exercise_normalized'] = gym['Exercise'].str.lower()

# # Define requirements from the assignment
# required_body_parts = [
#     "traps", "neck", "forearms", "abdominals", "biceps", "triceps", 
#     "chest", "back", "adductors", "abductors", "calves", 
#     "glutes", "hamstrings", "quadriceps"
# ]

# required_exercise_levels = ["beginner", "intermediate", "advanced"]
# required_categories = ["strongman", "powerlifting", "olympic weightlifting"]
# required_equipment = ["barbell", "dumbbell", "machine", "cable", "e-z curl bar", "band"]
# specific_exercises = ["barbell back squat"]  # As mentioned in question (f)

# # Normalize requirements for comparison
# required_terms = (
#     [item.lower() for item in required_body_parts] + 
#     [item.lower() for item in required_exercise_levels] + 
#     [item.lower() for item in required_categories] + 
#     [item.lower() for item in required_equipment] +
#     [item.lower() for item in specific_exercises]
# )

# # Identify matching rows
# matches = gym[gym['Exercise_normalized'].str.contains('|'.join(required_terms), case=False)]

# # Identify missing terms (those not found in the dataset)
# found_terms = set(matches['Exercise_normalized'])
# missing_terms = [term for term in required_terms if term not in found_terms]

# # Output results
# print(f"Matching Rows:\n{matches['Exercise'].values}")
# print(f"\nMissing Terms: {missing_terms}")

# # Save the matching rows to a CSV for reference
# matches.to_csv('/Users/Sam/Downloads/matching_exercises.csv', index=False)


* b

### Explanation for Question (b)

The objective of allocating **a proportion of the workout program to each exercise** is a more practical approach than specifying exact exercises for each session due to several reasons derived from the **assignment requirements** and the **structure of the provided dataset (`updated_gym_data.csv`)**:

---

#### **1. Flexibility in Program Design**
   - **Proportional Allocation**: By specifying proportions, the program provides flexibility in selecting exercises to meet diverse goals while adhering to the constraints. This allows customization for different lifters, whether they are beginners, intermediates, or advanced athletes.
   - **Dataset Variety**: The dataset contains 2,637 exercises across various body parts, levels, and equipment. Fixing exact exercises for each session would be overly restrictive and fail to utilize the dataset's diversity.

---

#### **2. Balancing Constraints Across Multiple Goals**
   - The constraints in the assignment (e.g., minimum proportions for specific body parts, SFR limits, and equipment requirements) are easier to satisfy with proportional allocations than with exact exercises.
   - For example:
     - **Body part constraints**: The requirement to allocate at least 2.5% to most body parts, with specific thresholds for traps, neck, forearms, and abdominals, can be evenly distributed across exercises targeting those muscles without overloading individual sessions.
     - **SFR Limit**: Keeping the overall program's stimulus-to-fatigue ratio below 0.55 ensures recovery is optimized without micromanaging fatigue per session.

---

#### **3. Adaptability to Time and Recovery**
   - **Time Management**: The dataset includes an estimate of the time required to complete 4 sets of 10 reps for each exercise. By allocating proportions, lifters can adjust the time spent on workouts to match their schedules.
   - **Recovery Optimization**: Proportional allocation allows the program to include lower-fatigue exercises for active recovery or balance high-fatigue exercises across sessions.

---

#### **4. Practicality for Diverse Populations**
   - A proportional program ensures inclusivity for lifters with different preferences or access to equipment:
     - Some may prefer barbell-based exercises, while others rely on machines or bands.
     - Not all gyms or home setups provide access to all equipment types.
   - By focusing on proportions, the program ensures that lifters can substitute exercises within the allocated categories while meeting the overall fitness goals.

---

#### **5. Alignment with Data Analytics**
   - The dataset provides detailed metrics (e.g., SFR, hypertrophy ratings, and estimated time). A proportional allocation allows optimization based on these metrics, creating a data-driven plan that maximizes hypertrophy while satisfying constraints.
   - Specifying exact exercises would not leverage the dataset's analytics capabilities, such as:
     - **Maximizing overall hypertrophy rating**: A proportional approach uses ratings across multiple exercises rather than focusing on single, potentially suboptimal choices.
     - **Meeting proportional constraints**: Constraints like 2.6x leg-to-upper-body allocation are better balanced with proportional distributions.

---

#### **6. Long-Term Sustainability**
   - Proportional allocation allows lifters to rotate or substitute exercises within the defined categories without disrupting the program's overall goals. This prevents overuse injuries, ensures variety, and keeps workouts engaging.
   - For instance:
     - Lifters can alternate between squats and lunges within the leg allocation or between barbell and dumbbell chest presses within the chest allocation.

---

#### Conclusion
Allocating a proportion of the workout program to each exercise leverages the richness of the dataset and the versatility of linear programming. It allows for a balanced, adaptable, and practical fitness program tailored to individual goals and constraints, avoiding the rigidity and impracticality of specifying exact exercises for every session.

* c

In [176]:
import pandas as pd
from gurobipy import Model, GRB, quicksum

# Load the dataset
gym = pd.read_csv('/Users/Sam/Downloads/updated_gym_data.csv')


In [177]:
# Extract relevant data for constraints
category = gym['Category'].tolist()
bodypart = gym['BodyPart'].tolist()
equipment = gym['Equipment'].fillna('None').tolist()
difficulty = gym['Difficulty'].tolist()
sfr = gym['Stimulus-to-Fatigue'].tolist()
hypertrophy = gym['Hypertrophy Rating'].tolist()


In [178]:
# Helper function to get indices for specific values in a column
def get_indices(values, target_list):
    return [i for i, value in enumerate(target_list) if value in values]

# Define key groups for constraints
legs = ['Adductors', 'Abductors', 'Calves', 'Glutes', 'Hamstrings', 'Quadriceps']
upper_body = ['Chest', 'Biceps', 'Triceps', 'Shoulders', 'Lats', 'Middle Back', 'Lower Back', 'Traps']
major_equipment = ['Barbell', 'Dumbbell', 'Machine', 'Cable', 'E-Z Curl Bar', 'Bands']
strongman_category = ['Strongman']
powerlifting_category = ['Powerlifting']
olympic_category = ['Olympic Weightlifting']

# Pre-compute indices for constraints
legs_indices = get_indices(legs, bodypart)
upper_body_indices = get_indices(upper_body, bodypart)
major_equipment_indices = get_indices(major_equipment, equipment)
strongman_indices = get_indices(strongman_category, category)
powerlifting_indices = get_indices(powerlifting_category, category)
olympic_indices = get_indices(olympic_category, category)
beginner_indices = get_indices(['Beginner'], difficulty)
intermediate_indices = get_indices(['Intermediate'], difficulty)
expert_indices = get_indices(['Expert'], difficulty)

# Back exercises specifically include "Lower Back" and "Middle Back"
back_exercises = ['Lower Back', 'Middle Back']
back_indices = get_indices(back_exercises, bodypart)
chest_indices = get_indices(['Chest'], bodypart)
biceps_indices = get_indices(['Biceps'], bodypart)
triceps_indices = get_indices(['Triceps'], bodypart)


In [179]:
# Initialize the Gurobi model
model = Model('WeightliftingOptimization')

# Define decision variables for exercise proportions
n_exercises = len(gym)
x = model.addVars(n_exercises, lb=0, ub=1, name="x")

# Constraint: Total proportions must sum to 1
model.addConstr(quicksum(x[i] for i in range(n_exercises)) == 1, "TotalProportion")


<gurobi.Constr *Awaiting Model Update*>

In [180]:
# Constraint 1: No single exercise > 5% of the program
for i in range(n_exercises):
    model.addConstr(x[i] <= 0.05, f"Max5Percent_{i}")

# Constraint 2: Minimum allocation for body parts
min_allocations = {
    'Traps': 0.005, 'Neck': 0.005, 'Forearms': 0.005, 'Abdominals': 0.04
}
for part, min_allocation in min_allocations.items():
    indices = get_indices([part], bodypart)
    model.addConstr(quicksum(x[i] for i in indices) >= min_allocation, f"MinAllocation_{part}")

for part in set(bodypart) - set(min_allocations.keys()):
    indices = get_indices([part], bodypart)
    model.addConstr(quicksum(x[i] for i in indices) >= 0.025, f"MinAllocation_{part}")

# Constraint 3: Leg muscles >= 2.6 × upper body muscles
model.addConstr(
    quicksum(x[i] for i in legs_indices) >= 2.6 * quicksum(x[i] for i in upper_body_indices), 
    "LegsVsUpperBody"
)

# Constraint 4: Equal proportions for Biceps & Triceps, Chest & All Back
model.addConstr(
    quicksum(x[i] for i in biceps_indices) == quicksum(x[i] for i in triceps_indices), 
    "BicepsTricepsEquality"
)
model.addConstr(
    quicksum(x[i] for i in chest_indices) == quicksum(x[i] for i in back_indices), 
    "ChestBackEquality"
)

# Constraint 5: Overall stimulus-to-fatigue ratio <= 0.55
model.addConstr(
    quicksum(x[i] * sfr[i] for i in range(n_exercises)) <= 0.55, 
    "MaxStimulusToFatigue"
)

# Constraint 6: Beginner >= 1.4 × Intermediate, Intermediate >= 1.1 × Expert
model.addConstr(
    quicksum(x[i] for i in beginner_indices) >= 1.4 * quicksum(x[i] for i in intermediate_indices), 
    "BeginnerVsIntermediate"
)
model.addConstr(
    quicksum(x[i] for i in intermediate_indices) >= 1.1 * quicksum(x[i] for i in expert_indices), 
    "IntermediateVsExpert"
)

# Constraint 7: Proportion constraints for exercise categories
model.addConstr(quicksum(x[i] for i in strongman_indices) <= 0.08, "MaxStrongman")
model.addConstr(quicksum(x[i] for i in powerlifting_indices) >= 0.09, "MinPowerlifting")
model.addConstr(quicksum(x[i] for i in olympic_indices) >= 0.10, "MinOlympic")

# Constraint 8: Proportion for major equipment >= 60%
model.addConstr(
    quicksum(x[i] for i in major_equipment_indices) >= 0.60, 
    "MajorEquipment"
)


<gurobi.Constr *Awaiting Model Update*>

In [181]:
# Objective: Maximize hypertrophy rating
model.setObjective(
    quicksum(x[i] * hypertrophy[i] for i in range(n_exercises)), 
    GRB.MAXIMIZE
)


In [182]:
# Optimize the model
model.optimize()

# Display results
if model.status == GRB.OPTIMAL:
    print(f"Optimal Hypertrophy Rating: {model.objVal:.4f}")
    for i in range(n_exercises):
        if x[i].x > 0:
            print(f"Exercise: {gym['Exercise'].iloc[i]}, Proportion: {x[i].x:.4f}")
else:
    print("No optimal solution found.")


Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 2665 rows, 2637 columns and 19569 nonzeros
Model fingerprint: 0x27879355
Coefficient statistics:
  Matrix range     [2e-01, 3e+00]
  Objective range  [3e-01, 1e+00]
  Bounds range     [1e+00, 1e+00]
  RHS range        [5e-03, 1e+00]
Presolve removed 2637 rows and 0 columns
Presolve time: 0.00s
Presolved: 28 rows, 2637 columns, 16932 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.0000000e+00   2.805000e+00   0.000000e+00      0s
      39    7.6649592e-01   0.000000e+00   0.000000e+00      0s

Solved in 39 iterations and 0.01 seconds (0.01 work units)
Optimal objective  7.664959245e-01
Optimal Hypertrophy Rating: 0.7665
Exercise: Clean Deadlift, Proportion: 0.0500
Exercise: Muscle Snatch, Proportion: 0.0500
Exercise: Clean Shrug, Proportion: 0.01

* d

In [183]:
# Adjusted SFR constraint with retrieval of variables
def adjust_sfr_constraint(model, sfr, relaxation=0.001):
    # Retrieve variables from the copied model
    x_copied = model.getVars()

    # Remove the old SFR constraint
    model.remove(model.getConstrByName("MaxStimulusToFatigue"))
    model.update()

    # Add the relaxed SFR constraint
    model.addConstr(
        quicksum(x_copied[i] * sfr[i] for i in range(len(x_copied))) <= 0.55 + relaxation,
        "MaxStimulusToFatigue"
    )
    model.update()
    return model



In [184]:
# Clone the model for the relaxed SFR scenario
model_relaxed = model.copy()

# Adjust the SFR constraint by 0.001
model_relaxed = adjust_sfr_constraint(model_relaxed, sfr, relaxation=0.001)

# Optimize the relaxed model
model_relaxed.optimize()

# Store the new optimal hypertrophy rating
if model_relaxed.status == GRB.OPTIMAL:
    relaxed_hypertrophy_rating = model_relaxed.objVal
    print(f"Relaxed Optimal Hypertrophy Rating: {relaxed_hypertrophy_rating:.4f}")
else:
    print("No optimal solution found for the relaxed model.")



Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 2665 rows, 2637 columns and 19569 nonzeros
Model fingerprint: 0x87a74bda
Coefficient statistics:
  Matrix range     [2e-01, 3e+00]
  Objective range  [3e-01, 1e+00]
  Bounds range     [1e+00, 1e+00]
  RHS range        [5e-03, 1e+00]
Presolve removed 2637 rows and 0 columns
Presolve time: 0.00s
Presolved: 28 rows, 2637 columns, 16932 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.0000000e+00   2.805000e+00   0.000000e+00      0s
      39    7.6787292e-01   0.000000e+00   0.000000e+00      0s

Solved in 39 iterations and 0.01 seconds (0.01 work units)
Optimal objective  7.678729187e-01
Relaxed Optimal Hypertrophy Rating: 0.7679


In [185]:
# Calculate the improvement in hypertrophy rating
if model.status == GRB.OPTIMAL and model_relaxed.status == GRB.OPTIMAL:
    original_hypertrophy_rating = model.objVal
    improvement = relaxed_hypertrophy_rating - original_hypertrophy_rating
    print(f"Hypertrophy Rating Improvement: {improvement:.4f}")

    # Evaluate validity of the estimate
    if improvement <= 0.001 * len(gym):
        print("The improvement estimate is valid since the relaxation is minor.")
    else:
        print("The improvement estimate might not be valid due to non-linear effects.")
else:
    print("One or both models did not solve optimally.")


Hypertrophy Rating Improvement: 0.0014
The improvement estimate is valid since the relaxation is minor.


* e

To answer Question (e), we will analyze whether increasing the minimum proportions for Traps, Neck, Forearms, or Abdominals would provide value, using the results from parts (b), (c), and (d). Here's the structured approach to address the question:

---

### Analysis Based on Results:

1. **Results from Question (c)**:
   - The optimal hypertrophy rating was **0.7665**, using all constraints.
   - Exercises targeting Traps, Neck, Forearms, and Abdominals were included in small proportions:
     - Lying Face Down Plate Neck Resistance (Neck): 0.0050
     - Rickshaw Carry (Traps): 0.0050
     - Crucifix (Forearms): 0.0250
     - 30 Barbell Floor Wiper (Abdominals): 0.0400

   These values indicate that the lower bounds set for these muscles were utilized, suggesting that their proportions are minimally sufficient to meet constraints but might not contribute much to the overall hypertrophy rating.

2. **Results from Question (d)**:
   - Relaxing the SFR constraint by 0.001 led to a **small improvement in hypertrophy rating (0.0014)**, resulting in a relaxed optimal hypertrophy rating of **0.7679**.
   - The improvement was minor and valid since the relaxation was small and the SFR constraint remained within reasonable bounds.

   This suggests that the hypertrophy rating is already close to being optimized, and significant improvements may not be achievable through minor changes in constraints.

3. **Insights from the Dataset**:
   - These muscles (Traps, Neck, Forearms, and Abdominals) often have lower hypertrophy ratings and stimulus-to-fatigue ratios compared to other major muscle groups (e.g., Quadriceps, Glutes).
   - Increasing their minimum proportions would likely reduce the overall hypertrophy rating because higher-proportion allocations would need to come from muscles or exercises with higher ratings.

---

### Conclusion for Question (e):
Increasing the minimum proportions for Traps, Neck, Forearms, or Abdominals is unlikely to provide significant value for the following reasons:

1. **Minimal Contribution to Hypertrophy Rating**:
   - The exercises targeting these muscles were included in the solution at the minimum required proportions, indicating their limited contribution to maximizing hypertrophy.

2. **Opportunity Cost**:
   - Increasing their proportions would require reducing allocations to other muscle groups (e.g., Legs or Chest), which generally have higher hypertrophy ratings and SFR values. This trade-off could decrease the overall hypertrophy rating.

3. **SFR and Hypertrophy Limitations**:
   - These muscles tend to have exercises with lower stimulus-to-fatigue ratios, meaning that their inclusion may disproportionately increase fatigue without significant hypertrophy benefits.

### Recommendation:
- There is little value in increasing the minimum proportions for Traps, Neck, Forearms, or Abdominals in the current model.
- However, if specific training goals (e.g., neck strengthening or grip improvement) are prioritized, increasing these proportions might be justified despite a potential decrease in the overall hypertrophy rating.

Let me know if you’d like me to refine this analysis further!

* f

In [192]:
# Identify the index of "Barbell Back Squats" in the dataset
barbell_back_squat_index = gym[gym['Exercise'] == "Barbell Back Squats"].index[0]
print(f"Index of 'Barbell Back Squats': {barbell_back_squat_index}")


Index of 'Barbell Back Squats': 331


In [193]:
from gurobipy import GRB

def find_required_hypertrophy_increase(model, x, gym, barbell_back_squat_index):
    # Original hypertrophy rating
    original_rating = gym.loc[barbell_back_squat_index, 'Hypertrophy Rating']
    print(f"Original Hypertrophy Rating: {original_rating}")

    # Initialize a small increment for testing
    increment = 0.001
    required_increase = 0

    while True:
        # Adjust hypertrophy rating
        new_rating = original_rating + required_increase
        hypertrophy[barbell_back_squat_index] = new_rating

        # Update the objective in the model
        model.setObjective(
            quicksum(x[i] * hypertrophy[i] for i in range(len(hypertrophy))), 
            GRB.MAXIMIZE
        )
        
        # Re-optimize the model
        model.optimize()

        # Check if "Barbell Back Squats" receives a positive allocation
        if model.status == GRB.OPTIMAL and x[barbell_back_squat_index].x > 0:
            print(f"'Barbell Back Squats' included with hypertrophy rating: {new_rating}")
            return new_rating - original_rating

        # Increment required increase
        required_increase += increment


In [194]:
required_increase = find_required_hypertrophy_increase(model, x, gym, barbell_back_squat_index)
print(f"Required Increase in Hypertrophy Rating for 'Barbell Back Squats': {required_increase:.4f}")


Original Hypertrophy Rating: 0.6783464
Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 2665 rows, 2637 columns and 19569 nonzeros
Coefficient statistics:
  Matrix range     [2e-01, 3e+00]
  Objective range  [3e-01, 1e+00]
  Bounds range     [1e+00, 1e+00]
  RHS range        [5e-03, 1e+00]
LP warm-start: use basis

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    7.6649592e-01   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.00 seconds (0.00 work units)
Optimal objective  7.664959245e-01
Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 2665 rows, 2637 columns and 19569 nonzeros
Coefficient statistics:
  Matrix range     [2e-01, 3e+00]

In [195]:
def find_required_hypertrophy_increase_optimized(model, x, gym, barbell_back_squat_index):
    original_rating = gym.loc[barbell_back_squat_index, 'Hypertrophy Rating']
    print(f"Original Hypertrophy Rating: {original_rating}")

    lower_bound = 0
    upper_bound = 1 - original_rating  # Assuming hypertrophy ratings must be <= 1
    tolerance = 1e-5
    required_increase = 0

    while upper_bound - lower_bound > tolerance:
        # Midpoint as the new test rating
        required_increase = (lower_bound + upper_bound) / 2
        new_rating = original_rating + required_increase
        hypertrophy[barbell_back_squat_index] = new_rating

        # Update objective
        model.setObjective(
            quicksum(x[i] * hypertrophy[i] for i in range(len(hypertrophy))), 
            GRB.MAXIMIZE
        )
        
        # Re-optimize
        model.optimize()

        # Check if Barbell Back Squats is included
        if model.status == GRB.OPTIMAL and x[barbell_back_squat_index].x > 0:
            upper_bound = required_increase  # Reduce the search range
        else:
            lower_bound = required_increase  # Increase the search range

    # Final required increase
    final_increase = (lower_bound + upper_bound) / 2
    print(f"'Barbell Back Squats' included with hypertrophy rating: {original_rating + final_increase:.4f}")
    return final_increase

# Call the optimized function
required_increase = find_required_hypertrophy_increase_optimized(model, x, gym, barbell_back_squat_index)
print(f"Required Increase in Hypertrophy Rating for 'Barbell Back Squats': {required_increase:.4f}")


Original Hypertrophy Rating: 0.6783464
Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 2665 rows, 2637 columns and 19569 nonzeros
Coefficient statistics:
  Matrix range     [2e-01, 3e+00]
  Objective range  [3e-01, 1e+00]
  Bounds range     [1e+00, 1e+00]
  RHS range        [5e-03, 1e+00]
LP warm-start: use basis

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    3.1842910e+01   2.622998e+03   0.000000e+00      0s
       7    7.6649592e-01   0.000000e+00   0.000000e+00      0s

Solved in 7 iterations and 0.00 seconds (0.00 work units)
Optimal objective  7.664959245e-01
Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (mac64[arm] - Darwin 24.2.0 24C101)

CPU model: Apple M3
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 2665 rows, 2637 columns and 19569 no